planner/cascades/transformation_rules.go

```diff
@@ -1448,3 +1452,174 @@ func (r *TransformLimitToTableDual) OnTransform(old *memo.ExprIter) (newExprs []
+	}
+	if contain {
+		newAggExpr := r.doEliminateOuterJoin(agg, join, joinExpr.Schema(), outerGroup, innerChildIdx)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+
+	return nil, false, false, nil
+}
+
+// It will transform `Aggregation->Join->outer,inner` to `Aggregation->outer` or `Aggregation->Selection->outer`
+func (r *EliminateOuterJoinBelowAggregation) doEliminateOuterJoin(agg *plannercore.LogicalAggregation, join *plannercore.LogicalJoin, joinSchema *expression.Schema, childGroup *memo.Group, innerChildIdx int) *memo.GroupExpr {
+	newAggExpr := memo.NewGroupExpr(agg)
+	ok, sel := convertOuterJoinToSel(join, innerChildIdx)
+	if ok {
+		selExpr := memo.NewGroupExpr(sel)
+		selExpr.SetChildren(childGroup)
+		selGroup := memo.NewGroupWithSchema(selExpr, joinSchema)
+		childGroup = selGroup
+	}
+	newAggExpr.SetChildren(childGroup)
+	return newAggExpr
+}
+
+// extract cols from agg
+func (r *EliminateOuterJoinBelowAggregation) getAggCols(agg *plannercore.LogicalAggregation) (aggCols []*expression.Column) {
+	for _, groupByItem := range agg.GroupByItems {
+		aggCols = append(aggCols, expression.ExtractColumns(groupByItem)...)
+	}
+	for _, aggDesc := range agg.AggFuncs {
+		for _, expr := range aggDesc.Args {
+			aggCols = append(aggCols, expression.ExtractColumns(expr)...)
+		}
+	}
+	return aggCols
+}
+
+// return true only all the aggregate functions are duplicate agnostic.
+// Only the following functions are considered to be duplicate agnostic:
+//   1. MAX(arg)
+//   2. MIN(arg)
+//   3. FIRST_ROW(arg)
+//   4. Other agg functions with DISTINCT flag, like SUM(DISTINCT arg)
+func (r *EliminateOuterJoinBelowAggregation) isAggFuncAllDupAgnostic(agg *plannercore.LogicalAggregation) bool {
+	for _, aggDesc := range agg.AggFuncs {
+		if !aggDesc.HasDistinct &&
+			aggDesc.Name != ast.AggFuncFirstRow &&
+			aggDesc.Name != ast.AggFuncMax &&
+			aggDesc.Name != ast.AggFuncMin {
+			return false
+		}
+	}
+	return true
+}
+
+// convert outer join to selection when len(outerConditions) > 0
+func convertOuterJoinToSel(join *plannercore.LogicalJoin, innerChildIdx int) (bool, *plannercore.LogicalSelection) {
+	var outerConditions expression.CNFExprs
+	switch innerChildIdx {
+	case 0:
+		outerConditions = join.RightConditions
+	case 1:
+		outerConditions = join.LeftConditions
+	default:
+		return false, nil
+	}
+
+	if len(outerConditions) > 0 {
+		newSel := plannercore.LogicalSelection{Conditions: outerConditions}.Init(join.SCtx(), join.SelectBlockOffset())
+		return true, newSel
+	}
+	return false, nil
+}
+
+// check whether the cols all from outer plan
+func isColsAllFromOuterTable(cols []*expression.Column, outerUniqueIDs set.Int64Set) bool {
+	for _, col := range cols {
+		if !outerUniqueIDs.Exist(col.UniqueID) {
+			return false
+		}
+	}
+	return true
+}
+
+// extract join keys as a schema for inner child of a outer join
+func extractInnerJoinKeys(join *plannercore.LogicalJoin, innerChildIdx int) *expression.Schema {
```

Comment:
Also, this method can be moved to `LogicalJoin.ExtractInnerJoinKeys`.

Reply:
For example， sql like `select * from a left join b on a.c = b.c and b.c is not null` can be converted to `select * from a join b on a.c = b.c`.
outer join can be converted to inner join .
How can I copy a `logical join` and only change the `join type`.
`logical join` has too many fields, and some fields is private.
Can I implement a function `copy` in `LogicalJoin` ?